### PR TITLE
[DNM][WIP] I am afraid I can't do that Dave. || AI Interaction Rework

### PR DESCRIPTION
**Pull request by Bibby0110**
### Intent of your Pull Request
Needs https://github.com/yogstation13/Yogstation/pull/8531 merged

The AI now needs processing relays to interact with machines.
These can be constructed, turned off, reconfigured and all the regular machine stuff.

Currently we have 5 types of relays.
Power Manipulation: APCs, Emitters, SMES units, 
Enviromental Control: Air Alarms, fire alarms, firelocks
Telecommunications Control: Using the radio, tcomms machinery.
Machine Interaction: Any other console/machine
Door Control: Opening/closing/whatever doors.

You can automatically connect to relays in your own area.
Other AIs can hack into remote relays that are not in their area by clicking them. THIS NOTIFIES ALL CONNECTED AI'S OF THE HACKING AI'S LOCATION

Relays can be purged from all connections (Even hacked ones) by hitting it with a multitool

**TODO:** 
**Maps**
- [ ] Update Box
- [ ] Update Delta
- [ ] Update Meta
- [ ] Update Omega
- [ ] Update Icebox

**Machines:**
- [x] Generate heat? Cool down AI core?
- [ ] Await sprites
- [ ] Make update_icon actually do something
- [x] All-in-one Malf Relay

#### Changelog

:cl:  
rscadd: AI Processing Relays
rscadd: AI's can no longer interact with items if the accompanying Processing Relay is offline/missing
/:cl:

### CODE REVIEW

**Comment by ToGWtF** [comment]
As the discussion for this took place on my discord server I claim partial if not full credit for this.

**Comment by JamieD1** [comment]
I concur with Tog, I take partial blame/credit for this.

**Comment by alexkar598** [comment]
So if an AI is malf it takes a person with a crowbar and a screwdriver and the AI is an expensive paperweight?

**Comment by Bibby0110** [comment]
> So if an AI is malf it takes a person with a crowbar and a screwdriver and the AI is an expensive paperweight?

Malf AI gets to build an all-in-one unit anywhere, just like the borging machine

**Comment by JamieD1** [comment]
> So if an AI is malf it takes a person with a crowbar and a screwdriver and the AI is an expensive paperweight?

What gamer said

**Comment by MegaEmpirical** [comment]
I'm curious about this and up for trying it out/expanding on it.

File: code/game/machinery/firealarm.dm [hunk 1]
```diff
@@ -148,6 +148,9 @@
 		alarm(user)
 
 /obj/machinery/firealarm/attack_ai(mob/user)
+	if(!ai_can_use(ENVIROMENTAL_CONTROL, user))
+		to_chat(user, "<span class='warning'>No Enviromental Control relay detected. Unable to interface with fire alarm.</span>")
```
Comment (Kmc2000) [inline code/game/machinery/firealarm.dm]: Having to rewrite this every single time is extremely inefficient and wasteful. Move this to "ai_can_use()"

File: code/game/machinery/ai_relay.dm [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		update_icon()
+
+
+/obj/machinery/ai_relay/process()
+	if(stat & (NOPOWER|BROKEN))
+		visible_message("[src] switches off.")
+		update_icon()
+		on = FALSE
+		use_power = IDLE_POWER_USE
+		return FALSE
```
Comment (Kmc2000) [inline code/game/machinery/ai_relay.dm:100]:
```suggestion
		return PROCESS_KILL
```
Then re-start the process when you need it?

File: code/game/machinery/ai_relay.dm [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	var/module_door_open = FALSE
+	var/obj/item/ai_relay_module/inserted_module
+
+	var/last_flipped
+
+	circuit = /obj/item/circuitboard/machine/ai_relay
+
+/obj/machinery/ai_relay/Initialize()
+	GLOB.ai_relays += src
+	name += " ([num2hex(rand(1,65535), -1)])"
```
Comment (Kmc2000) [inline code/game/machinery/ai_relay.dm]:
```suggestion
	name = "[name] ([num2hex(rand(1,65535), -1)])"
```

File: code/game/machinery/ai_relay.dm [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	critical_machine = TRUE
+	density = TRUE
+
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 500
+	active_power_usage = 5000 //10x
+
+	var/on = FALSE
+
+	var/list/linked_ais = list()
```
Comment (Kmc2000) [inline code/game/machinery/ai_relay.dm:22]: You could potentially optimise this to be a 2-d dictionary (k=String, v=list() ) like:

var/list/ais = list()

ais["hacked"] += a_hacked_ai_ref
ais["vanilla"] += an_ai_ref

**Review by Kmc2000 — CHANGES_REQUESTED** [review]

File: code/game/machinery/ai_relay.dm (hunk 2 above)
Comment (Bibby0110) [reply code/game/machinery/ai_relay.dm:100]: Nah we would need to run a process or similar  to detect the power change

Comment (Kmc2000) [reply code/game/machinery/ai_relay.dm:100]: I guess that works

File: code/game/machinery/firealarm.dm (hunk 1 above)
Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: Gah then we have to encode the name of the relay board somehow, got any nifty ideas?

File: code/game/machinery/ai_relay.dm (hunk 2 above)
Comment (Bibby0110) [reply code/game/machinery/ai_relay.dm:100]: What I’m saying is that it’s the same either way

File: code/game/machinery/ai_relay.dm (hunk 3 above)
Comment (Bibby0110) [reply code/game/machinery/ai_relay.dm]: Boringggg but sure

File: code/game/machinery/ai_relay.dm (hunk 4 above)
Comment (Bibby0110) [reply code/game/machinery/ai_relay.dm:22]: Good idea

File: code/game/machinery/firealarm.dm (hunk 1 above)
Comment (Kmc2000) [reply code/game/machinery/firealarm.dm]: #define ENVIRONMENTAL_CONTROL "environmental control"

Could work, then just use that string as your deciding factor for what relay means what. 

GLOBAL_LIST_EMPTY(relays)

GLOB.relays["environmental control] += src or whatever on the relay

Perhaps?

Comment (Kmc2000) [reply code/game/machinery/firealarm.dm]: (then on can_use)

if(GLOB.relays[control_type]?.on)
//Success
else
to_chat(user, "You need a [control_type] relay, online, to do that.")

Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: They’re bitflags so you can make dual relays.

Easiest and ugliest way would be to just hardcode some text

Comment (Kmc2000) [reply code/game/machinery/firealarm.dm]: Ooh fancy.

Hmm, then this becomes a bit harder. You could potentially keep the above idea but separate it by bitflag:

if(control_types & RELAY_DOORS)
GLOB.relays[RELAY_DOORS] += src

?

Then:

You need a [GLOB.relays[RELAY_DOORS]][1].name I suppose, but that sounds shitcodey the more I think about it 😕

Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: Can we get the value of a define if we have the string of the define name?

Comment (Kmc2000) [reply code/game/machinery/firealarm.dm]: DEFINE, to my knowledge, is just text replacement. The compiler changes all your instances of FOO into 1 if you #define FOO 1

Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: Yeah that’s what I thought

Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: Worst case we make a list with the bit flag values as keys and the string as the value? Can we do that?

Comment (Kmc2000) [reply code/game/machinery/firealarm.dm]: That doesn't sound too terrible tbh. You could also just lazycode it and go:

(can_ai_use)
if(interaction_type & ENVIRONMENTAL_CONTROL) to_chat(user, "you need EV control tbh")

Comment (Bibby0110) [reply code/game/machinery/firealarm.dm]: Yeah that’s the ultra lazy way

**Comment by MacHac** [comment]
It's an interesting mechanic.  I'm not really sure why it exists.

You claimed to me earlier that it wasn't really an AI nerf because these devices are in the AI core.  If that's the case, what's the point of adding something like this?  Feature bloat can be dangerous; that's what got us our telecomms system that only 5-10 people really understand.  If this isn't a nerf, you've introduced a whole new mechanic and system that won't ever really affect anything except to get the shuttle called faster when the AI survives a bombing.

If this is a nerf (and I think it probably is), you should think about adding a compensatory buff.  AIs are very easy to counter in any number of ways and are unique among potential antagonists in being unable to hide from security at all.  Moving the AI off the station may take the time-to-kill north of three minutes, but its survival ability when caught is still below any human crew.  Letting gamers neuter the AI with strategic bombings doesn't help.

You've clearly put a lot of work into this and I'm not against it in principle, but I don't really see this stuff interacting with the round constructively.

**Comment by MegaEmpirical** [comment]
People not knowing tcomms is because they're lazy. It's not hard.

**Comment by MacHac** [comment]
Or perhaps it's because it's an overcomplicated, brittle, sparsely documented system that doesn't interface well with anything else in the game, the details of which are only relevant to 2-3 jobs as 'advanced knowledge.'

If the players don't understand a mechanic, it's not the players' fault, it's the designer's.

**Comment by MegaEmpirical** [comment]
... If a player doesn't understand something because they're lazy it's obviously their fault.

There is plenty of info on the wiki. Could there be more? Sure, but everything you need is there and the little bit that isn't is easily picked up by anyone that toys with it.

**Comment by JamieD1** [comment]
> Or perhaps it's because it's an overcomplicated, brittle, sparsely documented system that doesn't interface well with anything else in the game, the details of which are only relevant to 2-3 jobs as 'advanced knowledge.'
> 
> If the players don't understand a mechanic, it's not the players' fault, it's the designer's.

If I dont understand rocket science is it sciences fault or mine for not learning and getting off my ass.

**Comment by patpol4** [comment]
I like it because it needs my pr merged

**Comment by JamieD1** [comment]
https://www.youtube.com/watch?v=5DX142NtqOE

**Comment by Bibby0110** [comment]
> It's an interesting mechanic. I'm not really sure why it exists.
> 
> You claimed to me earlier that it wasn't really an AI nerf because these devices are in the AI core. If that's the case, what's the point of adding something like this? Feature bloat can be dangerous; that's what got us our telecomms system that only 5-10 people really understand. If this isn't a nerf, you've introduced a whole new mechanic and system that won't ever really affect anything except to get the shuttle called faster when the AI survives a bombing.
> 
> If this is a nerf (and I think it probably is), you should think about adding a compensatory buff. AIs are very easy to counter in any number of ways and are unique among potential antagonists in being unable to hide from security at all. Moving the AI off the station may take the time-to-kill north of three minutes, but its survival ability when caught is still below any human crew. Letting gamers neuter the AI with strategic bombings doesn't help.
> 


Don't compare the tcomms system to this, thank you very much.
This is a nerf to building an AI in maint, correct. It's not much of a nerf to the round start AI. The difference between bombing the AI now and before isn't much bigger, other than it potentially creating unique and interesting situations where the AI is crippled in some instances.

AIs are already very powerful. They don't need additional buffs. If this was a giant nerf, like you make it out to be, then yes, they would warrant a buff.

> You've clearly put a lot of work into this and I'm not against it in principle, but I don't really see this stuff interacting with the round constructively.

You could say this for *literally* any other semi-advanced feature.

**Comment by boodaliboo** [comment]
fantastic pr

**Comment by MacHac** [comment]
Alright, so that's the use case I was interested in.  This PR exists to make it harder to build an AI outside of the core.  Me point is, this seems like a pretty complicated way to accomplish that specific thing.

You could remove the spare AI board from the RD's office and make it more expensive to print one.  That would also accomplish that specific thing without adding a 50-file independent system to the code.

I say it doesn't really interact with the round constructively because most rounds it won't interact with the round at all.   I expect it'll be pretty easy for an AI player to play the role for dozens of rounds without knowing what those machines even do.  There's no emergent behavior and no interaction with existing systems; in 99% of cases, they either sit in the core and have no impact on gameplay or get blown up seconds before the AI core does.

Furthermore, detaching these things from the AI core itself creates something of a rookie trap.  Building a secret AI in maint to accomplish your goals is exactly the sort of emergent shenanigans that makes rounds interesting, but needing to also build 5 more machines screws people that don't already know the system while only adding busywork for those that do.  If you can put together an AI, you probably have the resources and ability to build a few more machines.

If these machines provided an advantage when upgraded or had any deeper strategy than 'on good off bad', my opinion might change.  As it stands, I think there are better ways to do what you're trying to do.

**Comment by Bibby0110** [comment]
> Alright, so that's the use case I was interested in. This PR exists to make it harder to build an AI outside of the core. Me point is, this seems like a pretty complicated way to accomplish that specific thing.
> 
> You could remove the spare AI board from the RD's office and make it more expensive to print one. That would also accomplish that specific thing without adding a 50-file independent system to the code.
> 
> I say it doesn't really interact with the round constructively because most rounds it won't interact with the round at all. I expect it'll be pretty easy for an AI player to play the role for dozens of rounds without knowing what those machines even do. There's no emergent behavior and no interaction with existing systems; in 99% of cases, they either sit in the core and have no impact on gameplay or get blown up seconds before the AI core does.
> 
> Furthermore, detaching these things from the AI core itself creates something of a rookie trap. Building a secret AI in maint to accomplish your goals is exactly the sort of emergent shenanigans that makes rounds interesting, but needing to also build 5 more machines screws people that don't already know the system while only adding busywork for those that do. If you can put together an AI, you probably have the resources and ability to build a few more machines.
> 
> If these machines provided an advantage when upgraded or had any deeper strategy than 'on good off bad', my opinion might change. As it stands, I think there are better ways to do what you're trying to do.

I'm under the impression that you haven't actually read and understood the PR.

If you did, you would understand that it's not a "rookie trap" but an alternative way to nerf maint AI's, while still keeping them semi-viable.

You either go loud with your AI and hack the normal relays, or you build a cozy little AI core in maint.

Just because something isn't drawn into every single round does not make it "bad" or "feature-bloat" as you say. 
It adds a nice possiblity for new situations and new RP.
Shutting/neutering the AI like this is much more interactive and fun than "laser hit AI go brrr!!"

I fail to see your argument, either because it's not there, or because you're misunderstanding *what* this is and how it's intended to be used.

**Comment by MacHac** [comment]
I just think it's overkill for what you're trying to do, and that your time and effort would be better served tweaking existing systems rather than creating a new one.  The code bloat and learning curve are problems, but I don't think it's gonna affect that much either way.

People being able to neuter the AI without killing it is a cool possibility, but it's going to depend heavily on map design.  It'll be interesting to see if it works.

I withdraw my objection.

**Comment by JamieD1** [comment]
> I just think it's overkill for what you're trying to do, and that your time and effort would be better served tweaking existing systems rather than creating a new one. The code bloat and learning curve are problems, but I don't think it's gonna affect that much either way.
> 
> People being able to neuter the AI without killing it is a cool possibility, but it's going to depend heavily on map design. It'll be interesting to see if it works.
> 
> I withdraw my objection.

Hes not trying to nerve maint ai, we could do that several different ways, this is a new ai system to bring more variety and options to how rounds can play out.

**Comment by Bibby0110** [comment]
> I just think it's overkill for what you're trying to do, and that your time and effort would be better served tweaking existing systems rather than creating a new one. The code bloat and learning curve are problems, but I don't think it's gonna affect that much either way.
> 
> People being able to neuter the AI without killing it is a cool possibility, but it's going to depend heavily on map design. It'll be interesting to see if it works.
> 
> I withdraw my objection.

There's barely a learning curve here. It's 1 or 2 things extra to keep in mind when things go south.

The AI core is recieving a makeover, yes.

File: code/game/machinery/ai_relay.dm (hunk 4 above)
Comment (Bibby0110) [reply code/game/machinery/ai_relay.dm:22]: I'm ditching this idea, it looked good on the surface, but it's not very practical

**Comment by alexkar598** [comment]
stale and abadonned AFAIK. Contact me on discord if you wish to reoopen this pr

**Comment by alexkar598** [comment]
Bibby has been teasing us with unpushed commits

**Comment by alexkar598** [comment]
It has been two fucking months. Im willing to let draft PRs live for a while but theres a limit on how far a PR can clog up the PR list without any **visible** progress.

**Comment by MegaEmpirical** [comment]
Sad.

**Pull request closed without merging**